In [31]:
# Import all required dependencies

import pandas as pd
import numpy as np
from matplotlib import pyplot as plt

from sqlalchemy import create_engine
import psycopg2
from config import db_password


### Import the new_selection.csv 

The `new_selection.csv` has data about all the soccer matches.

In [2]:
# file directory
file_dir = "C:/Users/mamoo/UT_DataScience/UTA-VIRT-DATA-PT-02-2021-U-B/FinalProject/Segment_2/"

# import the all_player.csv
mls_df = pd.read_csv(f"{file_dir}new_selection.csv", encoding= 'unicode_escape')
pd.set_option("display.max_columns", None) # To show all columns explicitly
mls_df

Unnamed: 0      id                  home                    away  \
0           1797  336076       Colorado Rapids        Columbus Crew SC   
1           1798  336077   Vancouver Whitecaps         Montreal Impact   
2           1799  336078             DC United    Sporting Kansas City   
3           1800  336080             LA Galaxy          Real Salt Lake   
4           1801  336079  San Jose Earthquakes  New England Revolution   
...          ...     ...                   ...                     ...   
3429        5226  597845   Vancouver Whitecaps          Real Salt Lake   
3430        5227  597846             LA Galaxy               FC Dallas   
3431        5228  597848          Nashville SC       Atlanta United FC   
3432        5229  597847    New York Red Bulls      Philadelphia Union   
3433        5230  597840         FC Cincinnati           Columbus Crew   

                    date  year attendance  \
0     Saturday, March 10  2012     14,746   
1     Saturday, March 10  2012     21,000   
2     Saturday, March 10  2012     16,314   
3     Saturday, March 10  2012     27,000   
4     Saturday, March 10  2012     10,525   
...                  ...   ...        ...   
3429   Wednesday, July 7  2021     12,009   
3430   Wednesday, July 7  2021     16,453   
3431    Thursday, July 8  2021     22,913   
3432    Thursday, July 8  2021     10,013   
3433      Friday, July 9  2021     25,701   

                                         venue   part_of_competition  \
0           Dick's Sporting Goods Park, Denver   Regular Season 2012   
1                          BC Place, Vancouver   Regular Season 2012   
2                RFK Stadium, Washington, D.C.   Regular Season 2012   
3      Dignity Health Sports Park, Los Angeles   Regular Season 2012   
4                            Buck Shaw Stadium   Regular Season 2012   
...                                        ...                   ...   
3429              Rio Tinto Stadium, Sandy, UT        Regular Season   
3430        Dignity Health Sports Park, Carson        Regular Season   
3431  Nissan Stadium, Nashville, Nashville, TN        Regular Season   
3432                  Red Bull Arena, Harrison        Regular Season   
3433               TQL Stadium, Cincinnati, OH        Regular Season   

     game_status  shootout  home_score  away_score home_possessionPct  \
0             FT     False           2           0                53%   
1             FT     False           2           0                50%   
2             FT     False           0           1                43%   
3             FT     False           1           3                49%   
4             FT     False           1           0                52%   
...          ...       ...         ...         ...                ...   
3429          FT     False           0           4                55%   
3430          FT     False           3           1                53%   
3431          FT     False           2           2                50%   
3432          FT     False           1           1                47%   
3433          FT     False           2           2                50%   

     away_possessionPct home_shotsSummary away_shotsSummary  \
0                   47%            17 (5)             5 (3)   
1                   50%            11 (5)            15 (7)   
2                   57%             7 (1)            17 (8)   
3                   51%            17 (4)            13 (4)   
4                   48%             5 (1)             6 (1)   
...                 ...               ...               ...   
3429                45%            10 (2)            18 (8)   
3430                47%            12 (5)            13 (5)   
3431                50%            17 (8)             6 (2)   
3432                53%            14 (3)            11 (7)   
3433                50%            13 (3)             9 (2)   

      home_foulsCommitted  away_foulsCommitted  home_yellowCards  \
0        

In [3]:
# Check all the column headers.
mls_df.columns

Index(['Unnamed: 0', 'id', 'home', 'away', 'date', 'year', 'attendance',
       'venue', 'part_of_competition', 'game_status', 'shootout', 'home_score',
       'away_score', 'home_possessionPct', 'away_possessionPct',
       'home_shotsSummary', 'away_shotsSummary', 'home_foulsCommitted',
       'away_foulsCommitted', 'home_yellowCards', 'away_yellowCards',
       'home_redCards', 'away_redCards', 'home_offsides', 'away_offsides',
       'home_wonCorners', 'away_wonCorners', 'home_saves', 'away_saves',
       'home_formation', 'away_formation', 'Outcome'],
      dtype='object')

In [4]:
mls_df.isnull().sum()

Unnamed: 0               0
id                       0
home                     0
away                     0
date                     0
year                     0
attendance             314
venue                   15
part_of_competition      0
game_status              0
shootout                 0
home_score               0
away_score               0
home_possessionPct       0
away_possessionPct       0
home_shotsSummary        0
away_shotsSummary        0
home_foulsCommitted      0
away_foulsCommitted      0
home_yellowCards         0
away_yellowCards         0
home_redCards            0
away_redCards            0
home_offsides            0
away_offsides            0
home_wonCorners          0
away_wonCorners          0
home_saves               0
away_saves               0
home_formation           0
away_formation           0
Outcome                  0
dtype: int64

It looks like `attendance` column has 314 null values which is less than 9% of total data rows. We may **drop** those rows with null values.

In [5]:
mls_df = mls_df.drop(mls_df[mls_df['attendance'].isnull()].index)
mls_df

Unnamed: 0      id                  home                    away  \
0           1797  336076       Colorado Rapids        Columbus Crew SC   
1           1798  336077   Vancouver Whitecaps         Montreal Impact   
2           1799  336078             DC United    Sporting Kansas City   
3           1800  336080             LA Galaxy          Real Salt Lake   
4           1801  336079  San Jose Earthquakes  New England Revolution   
...          ...     ...                   ...                     ...   
3429        5226  597845   Vancouver Whitecaps          Real Salt Lake   
3430        5227  597846             LA Galaxy               FC Dallas   
3431        5228  597848          Nashville SC       Atlanta United FC   
3432        5229  597847    New York Red Bulls      Philadelphia Union   
3433        5230  597840         FC Cincinnati           Columbus Crew   

                    date  year attendance  \
0     Saturday, March 10  2012     14,746   
1     Saturday, March 10  2012     21,000   
2     Saturday, March 10  2012     16,314   
3     Saturday, March 10  2012     27,000   
4     Saturday, March 10  2012     10,525   
...                  ...   ...        ...   
3429   Wednesday, July 7  2021     12,009   
3430   Wednesday, July 7  2021     16,453   
3431    Thursday, July 8  2021     22,913   
3432    Thursday, July 8  2021     10,013   
3433      Friday, July 9  2021     25,701   

                                         venue   part_of_competition  \
0           Dick's Sporting Goods Park, Denver   Regular Season 2012   
1                          BC Place, Vancouver   Regular Season 2012   
2                RFK Stadium, Washington, D.C.   Regular Season 2012   
3      Dignity Health Sports Park, Los Angeles   Regular Season 2012   
4                            Buck Shaw Stadium   Regular Season 2012   
...                                        ...                   ...   
3429              Rio Tinto Stadium, Sandy, UT        Regular Season   
3430        Dignity Health Sports Park, Carson        Regular Season   
3431  Nissan Stadium, Nashville, Nashville, TN        Regular Season   
3432                  Red Bull Arena, Harrison        Regular Season   
3433               TQL Stadium, Cincinnati, OH        Regular Season   

     game_status  shootout  home_score  away_score home_possessionPct  \
0             FT     False           2           0                53%   
1             FT     False           2           0                50%   
2             FT     False           0           1                43%   
3             FT     False           1           3                49%   
4             FT     False           1           0                52%   
...          ...       ...         ...         ...                ...   
3429          FT     False           0           4                55%   
3430          FT     False           3           1                53%   
3431          FT     False           2           2                50%   
3432          FT     False           1           1                47%   
3433          FT     False           2           2                50%   

     away_possessionPct home_shotsSummary away_shotsSummary  \
0                   47%            17 (5)             5 (3)   
1                   50%            11 (5)            15 (7)   
2                   57%             7 (1)            17 (8)   
3                   51%            17 (4)            13 (4)   
4                   48%             5 (1)             6 (1)   
...                 ...               ...               ...   
3429                45%            10 (2)            18 (8)   
3430                47%            12 (5)            13 (5)   
3431                50%            17 (8)             6 (2)   
3432                53%            14 (3)            11 (7)   
3433                50%            13 (3)             9 (2)   

      home_foulsCommitted  away_foulsCommitted  home_yellowCards  \
0        

In [6]:
# Check the "mls_df" datatypes.
mls_df.dtypes

Unnamed: 0              int64
id                      int64
home                   object
away                   object
date                   object
year                    int64
attendance             object
venue                  object
part_of_competition    object
game_status            object
shootout                 bool
home_score              int64
away_score              int64
home_possessionPct     object
away_possessionPct     object
home_shotsSummary      object
away_shotsSummary      object
home_foulsCommitted     int64
away_foulsCommitted     int64
home_yellowCards        int64
away_yellowCards        int64
home_redCards           int64
away_redCards           int64
home_offsides           int64
away_offsides           int64
home_wonCorners         int64
away_wonCorners         int64
home_saves              int64
away_saves              int64
home_formation         object
away_formation         object
Outcome                object
dtype: object

There are few columns having `object` datatype with inconsistent data values. Some of those columns seem to have no or low impact on decision making by the neural network model. We may choose to delete some column as shown below:

In [7]:
# Drop the columns 

mls_df = mls_df.drop("part_of_competition", axis='columns')
mls_df = mls_df.drop("home_shotsSummary", axis='columns')
mls_df = mls_df.drop("away_shotsSummary", axis='columns')
mls_df = mls_df.drop("home_formation", axis='columns')
mls_df = mls_df.drop("away_formation", axis='columns')
mls_df = mls_df.drop("date", axis='columns')

mls_df

Unnamed: 0      id                  home                    away  year  \
0           1797  336076       Colorado Rapids        Columbus Crew SC  2012   
1           1798  336077   Vancouver Whitecaps         Montreal Impact  2012   
2           1799  336078             DC United    Sporting Kansas City  2012   
3           1800  336080             LA Galaxy          Real Salt Lake  2012   
4           1801  336079  San Jose Earthquakes  New England Revolution  2012   
...          ...     ...                   ...                     ...   ...   
3429        5226  597845   Vancouver Whitecaps          Real Salt Lake  2021   
3430        5227  597846             LA Galaxy               FC Dallas  2021   
3431        5228  597848          Nashville SC       Atlanta United FC  2021   
3432        5229  597847    New York Red Bulls      Philadelphia Union  2021   
3433        5230  597840         FC Cincinnati           Columbus Crew  2021   

     attendance                                     venue game_status  \
0        14,746        Dick's Sporting Goods Park, Denver          FT   
1        21,000                       BC Place, Vancouver          FT   
2        16,314             RFK Stadium, Washington, D.C.          FT   
3        27,000   Dignity Health Sports Park, Los Angeles          FT   
4        10,525                         Buck Shaw Stadium          FT   
...         ...                                       ...         ...   
3429     12,009              Rio Tinto Stadium, Sandy, UT          FT   
3430     16,453        Dignity Health Sports Park, Carson          FT   
3431     22,913  Nissan Stadium, Nashville, Nashville, TN          FT   
3432     10,013                  Red Bull Arena, Harrison          FT   
3433     25,701               TQL Stadium, Cincinnati, OH          FT   

      shootout  home_score  away_score home_possessionPct away_possessionPct  \
0        False           2           0                53%                47%   
1        False           2           0                50%                50%   
2        False           0           1                43%                57%   
3        False           1           3                49%                51%   
4        False           1           0                52%                48%   
...        ...         ...         ...                ...                ...   
3429     False           0           4                55%                45%   
3430     False           3           1                53%                47%   
3431     False           2           2                50%                50%   
3432     False           1           1                47%                53%   
3433     False           2           2                50%                50%   

      home_foulsCommitted  away_foulsCommitted  home_yellowCards  \
0                      11                   14                 1   
1                      13                   26                 1   
2                      12                   16                 1   
3                      14                   10                 0   
4                      17                   14                 2   
...                   ...                  ...               ...   
3429                    7                    9                 0   
3430                    7                   15                 1   
3431                   16                   17                 3   
3432                   14                    9                 2   
3433                   17                   15                 6   

      away_yellowCards  home_redCards  away_redCards  home_offsides  \
0                    3              0              0              2   
1                    2              0              0              2   
2                    1              0              0              3   
3                    2              0              0              3   
4                    1             

In [8]:
# Now check again if there is any Null values.
mls_df.isnull().sum()

Unnamed: 0             0
id                     0
home                   0
away                   0
year                   0
attendance             0
venue                  0
game_status            0
shootout               0
home_score             0
away_score             0
home_possessionPct     0
away_possessionPct     0
home_foulsCommitted    0
away_foulsCommitted    0
home_yellowCards       0
away_yellowCards       0
home_redCards          0
away_redCards          0
home_offsides          0
away_offsides          0
home_wonCorners        0
away_wonCorners        0
home_saves             0
away_saves             0
Outcome                0
dtype: int64

So, there is **no more NULL** values.

In [9]:
# Check the "mls_df" datatypes.
mls_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3120 entries, 0 to 3433
Data columns (total 26 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Unnamed: 0           3120 non-null   int64 
 1   id                   3120 non-null   int64 
 2   home                 3120 non-null   object
 3   away                 3120 non-null   object
 4   year                 3120 non-null   int64 
 5   attendance           3120 non-null   object
 6   venue                3120 non-null   object
 7   game_status          3120 non-null   object
 8   shootout             3120 non-null   bool  
 9   home_score           3120 non-null   int64 
 10  away_score           3120 non-null   int64 
 11  home_possessionPct   3120 non-null   object
 12  away_possessionPct   3120 non-null   object
 13  home_foulsCommitted  3120 non-null   int64 
 14  away_foulsCommitted  3120 non-null   int64 
 15  home_yellowCards     3120 non-null   int64 
 16  away_y

It looks like `attendance`, `home_possessionPct`, and `away_possessionPct` columns are shown as type `object` which should not be. 

Let's check the `attendance` column, it has **comma ','** as thousand separator which should be removed.

The `home_possessionPct` and `away_possessionPct` columns has values in **percent (%)** format, which needs to be formatted as float or integer.


In [10]:
# Replace the ',' from attendance values.
mls_df['attendance'] = mls_df['attendance'].str.replace(",", "").astype("float")
mls_df.head(2)

Unnamed: 0      id                 home              away  year  \
0        1797  336076      Colorado Rapids  Columbus Crew SC  2012   
1        1798  336077  Vancouver Whitecaps   Montreal Impact  2012   

   attendance                               venue game_status  shootout  \
0     14746.0  Dick's Sporting Goods Park, Denver          FT     False   
1     21000.0                 BC Place, Vancouver          FT     False   

   home_score  away_score home_possessionPct away_possessionPct  \
0           2           0                53%                47%   
1           2           0                50%                50%   

   home_foulsCommitted  away_foulsCommitted  home_yellowCards  \
0                   11                   14                 1   
1                   13                   26                 1   

   away_yellowCards  home_redCards  away_redCards  home_offsides  \
0                 3              0              0              2   
1                 2              0              0              2   

   away_offsides  home_wonCorners  away_wonCorners  home_saves  away_saves  \
0              4                8                4           3           3   
1              2                4                4           7           3   

  Outcome  
0     Win  
1     Win

In [11]:
# Convert "homepossessionPct" % to float
mls_df['home_possessionPct'] = mls_df['home_possessionPct'].str.rstrip('%').astype('float') / 100.0
mls_df['away_possessionPct'] = mls_df['away_possessionPct'].str.rstrip('%').astype('float') / 100.0
#                                                     ^ use str func to elim '%'     ^ divide by 100

mls_df

Unnamed: 0      id                  home                    away  year  \
0           1797  336076       Colorado Rapids        Columbus Crew SC  2012   
1           1798  336077   Vancouver Whitecaps         Montreal Impact  2012   
2           1799  336078             DC United    Sporting Kansas City  2012   
3           1800  336080             LA Galaxy          Real Salt Lake  2012   
4           1801  336079  San Jose Earthquakes  New England Revolution  2012   
...          ...     ...                   ...                     ...   ...   
3429        5226  597845   Vancouver Whitecaps          Real Salt Lake  2021   
3430        5227  597846             LA Galaxy               FC Dallas  2021   
3431        5228  597848          Nashville SC       Atlanta United FC  2021   
3432        5229  597847    New York Red Bulls      Philadelphia Union  2021   
3433        5230  597840         FC Cincinnati           Columbus Crew  2021   

      attendance                                     venue game_status  \
0        14746.0        Dick's Sporting Goods Park, Denver          FT   
1        21000.0                       BC Place, Vancouver          FT   
2        16314.0             RFK Stadium, Washington, D.C.          FT   
3        27000.0   Dignity Health Sports Park, Los Angeles          FT   
4        10525.0                         Buck Shaw Stadium          FT   
...          ...                                       ...         ...   
3429     12009.0              Rio Tinto Stadium, Sandy, UT          FT   
3430     16453.0        Dignity Health Sports Park, Carson          FT   
3431     22913.0  Nissan Stadium, Nashville, Nashville, TN          FT   
3432     10013.0                  Red Bull Arena, Harrison          FT   
3433     25701.0               TQL Stadium, Cincinnati, OH          FT   

      shootout  home_score  away_score  home_possessionPct  \
0        False           2           0                0.53   
1        False           2           0                0.50   
2        False           0           1                0.43   
3        False           1           3                0.49   
4        False           1           0                0.52   
...        ...         ...         ...                 ...   
3429     False           0           4                0.55   
3430     False           3           1                0.53   
3431     False           2           2                0.50   
3432     False           1           1                0.47   
3433     False           2           2                0.50   

      away_possessionPct  home_foulsCommitted  away_foulsCommitted  \
0                   0.47                   11                   14   
1                   0.50                   13                   26   
2                   0.57                   12                   16   
3                   0.51                   14                   10   
4                   0.48                   17                   14   
...                  ...                  ...                  ...   
3429                0.45                    7                    9   
3430                0.47                    7                   15   
3431                0.50                   16                   17   
3432                0.53                   14                    9   
3433                0.50                   17                   15   

      home_yellowCards  away_yellowCards  home_redCards  away_redCards  \
0                    1                 3              0              0   
1                    1                 2              0              0   
2                    1                 1              0              0   
3                    0                 2              0              0   
4                    2                 1              0              0   
...                ...               ...            ...            ...   
3429                 0                 2            

In [12]:
mls_df.describe()

Unnamed: 0             id         year    attendance   home_score  \
count  3120.000000    3120.000000  3120.000000   3120.000000  3120.000000   
mean   3374.733654  445397.221154  2015.973718  20487.588782     1.714423   
std     925.449778   73154.817871     2.519297   9296.708746     1.261122   
min    1797.000000  336076.000000  2012.000000    222.000000     0.000000   
25%    2580.750000  384712.750000  2014.000000  15847.750000     1.000000   
50%    3363.500000  440812.500000  2016.000000  19100.000000     2.000000   
75%    4143.250000  502583.250000  2018.000000  22067.500000     2.000000   
max    5230.000000  597856.000000  2021.000000  73019.000000     7.000000   

        away_score  home_possessionPct  away_possessionPct  \
count  3120.000000         3120.000000         3120.000000   
mean      1.141346            0.522631            0.477048   
std       1.078019            0.083230            0.083143   
min       0.000000            0.000000            0.000000   
25%       0.000000            0.470000            0.420000   
50%       1.000000            0.520000            0.480000   
75%       2.000000            0.580000            0.530000   
max       7.000000            0.830000            0.780000   

       home_foulsCommitted  away_foulsCommitted  home_yellowCards  \
count          3120.000000          3120.000000       3120.000000   
mean             12.076603            12.483654          1.548718   
std               3.710020             3.825597          1.168217   
min               0.000000             0.000000          0.000000   
25%               9.000000            10.000000          1.000000   
50%              12.000000            12.000000          1.000000   
75%              14.000000            15.000000          2.000000   
max              28.000000            28.000000          6.000000   

       away_yellowCards  home_redCards  away_redCards  home_offsides  \
count       3120.000000    3120.000000    3120.000000    3120.000000   
mean           1.793590       0.086538       0.128526       2.170192   
std            1.183009       0.297814       0.365848       1.730720   
min            0.000000       0.000000       0.000000       0.000000   
25%            1.000000       0.000000       0.000000       1.000000   
50%            2.000000       0.000000       0.000000       2.000000   
75%            3.000000       0.000000       0.000000       3.000000   
max            7.000000       3.000000       3.000000      12.000000   

       away_offsides  home_wonCorners  away_wonCorners   home_saves  \
count    3120.000000      3120.000000      3120.000000  3120.000000   
mean        1.926603         5.746154         4.293590     2.786538   
std         1.660336         2.876141         2.480903     1.799846   
min         0.000000         0.000000         0.000000     0.000000   
25%         1.000000         4.000000         3.000000     1.000000   
50%         2.000000         5.000000         4.000000     3.000000   
75%         3.000000         8.000000         6.000000     4.000000   
max        14.000000        17.000000        17.000000    14.000000   

        away_saves  
count  3120.000000  
mean      3.516667  
std       2.059015  
min       0.000000  
25%       2.000000  
50%       3.000000  
75%       5.000000  
max      16.000000

In [23]:
# take a look at the 'home' column for all the team/club
mls_df['home'].unique()

array(['Colorado Rapids', 'Vancouver Whitecaps', 'DC United', 'LA Galaxy',
       'San Jose Earthquakes', 'FC Dallas', 'Chivas USA',
       'Portland Timbers', 'Montreal Impact', 'Sporting Kansas City',
       'Real Salt Lake', 'Seattle Sounders FC', 'Philadelphia Union',
       'Toronto FC', 'Columbus Crew SC', 'New England Revolution',
       'Chicago Fire FC', 'New York Red Bulls', 'Houston Dynamo',
       'Orlando City SC', 'New York City FC', 'Atlanta United FC',
       'Minnesota United FC', 'LAFC', 'FC Cincinnati',
       'Houston Dynamo FC', 'Nashville SC', 'Inter Miami CF',
       'Columbus Crew', 'Austin FC'], dtype=object)

In [26]:
# Check for the 'Austin FC' club
mls_df[mls_df['home'] == 'Austin FC']

Unnamed: 0      id       home              away  year  attendance  \
3408        5205  597823  Austin FC     Columbus Crew  2021     20738.0   
3410        5207  597832  Austin FC  Portland Timbers  2021     20738.0   
3426        5223  597842  Austin FC              LAFC  2021     20738.0   

                   venue game_status  shootout  home_score  away_score  \
3408  Q2 Stadium, Austin          FT     False           0           0   
3410  Q2 Stadium, Austin          FT     False           4           1   
3426  Q2 Stadium, Austin          FT     False           0           2   

      home_possessionPct  away_possessionPct  home_foulsCommitted  \
3408                0.57                0.43                   13   
3410                0.69                0.31                   11   
3426                0.59                0.41                    7   

      away_foulsCommitted  home_yellowCards  away_yellowCards  home_redCards  \
3408                    7                 1                 1              0   
3410                   10                 2                 3              0   
3426                    9                 3                 2              0   

      away_redCards  home_offsides  away_offsides  home_wonCorners  \
3408              0              0              2                6   
3410              0              0              1               13   
3426              0              2              3                8   

      away_wonCorners  home_saves  away_saves Outcome  
3408                6           6           5     Tie  
3410                4           0           3     Win  
3426                3           4           4    Loss

The **mls_df** is now **cleaned** with total of **26 columns, 3120 rows** of non-null data with consistent data types. 

### Export the clean df to .csv

In [13]:
# Export the clean dataframe to "matches_cleaned.csv"
csv_dir = "C:/Users/mamoo/UT_DataScience/UTA-VIRT-DATA-PT-02-2021-U-B/FinalProject/Resources/Clean_csv/"
mls_df.to_csv(f"{csv_dir}matches_cleaned.csv")

Since we're trying to predict the win/loss outcome of a **home** team, it'd be better to make another dataframe and .csv by applying `groupby` on `home` and `year` columns.

In [15]:
# Group by "home" and "Year"
matches_grouped = mls_df.groupby(['home','year']).sum()
matches_grouped

Unnamed: 0        id  attendance  shootout  \
home                year                                               
Atlanta United FC   2017       67556   8512648    886625.0         1   
                    2018       82495  10129029   1114594.0         0   
                    2019       91151  10738150   1044339.0         0   
                    2020       14787   1731632     81580.0         0   
                    2021       25759   2987797    186297.0         0   
...                              ...       ...         ...       ...   
Vancouver Whitecaps 2017       70996   9011030    412993.0         0   
                    2018       69870   8543381    373089.0         0   
                    2019       76836   9063761    331745.0         0   
                    2020        4752    560545     22120.0         0   
                    2021        5226    597845     12009.0         0   

                          home_score  away_score  home_possessionPct  \
home                year                                               
Atlanta United FC   2017          48          19               11.10   
                    2018          50          20               11.01   
                    2019          40          16               11.87   
                    2020           5           3                1.58   
                    2021           8           5                3.06   
...                              ...         ...                 ...   
Vancouver Whitecaps 2017          35          20                8.18   
                    2018          32          29                8.24   
                    2019          16          23                7.86   
                    2020           1           3                0.48   
                    2021           0           4                0.55   

                          away_possessionPct  home_foulsCommitted  \
home                year                                            
Atlanta United FC   2017                6.90                  214   
                    2018                8.99                  205   
                    2019                8.13                  205   
                    2020                1.42                   40   
                    2021                1.94                   55   
...                                      ...                  ...   
Vancouver Whitecaps 2017               10.82                  245   
                    2018                8.76                  205   
                    2019                9.14                  190   
                    2020                0.52                   10   
                    2021                0.45                    7   

                          away_foulsCommitted  home_yellowCards  \
home                year                                          
Atlanta United FC   2017                  211                25   
                    2018                  239                32   
                    2019                  262                28   
                    2020                   42                 4   
                    2021                   91                 8   
...                                       ...               ...   
Vancouver Whitecaps 2017                  217                30   
                    2018                  190                23   
                    2019                  209                37   
                    2020                   13                 1   
                    2021                    9                 0   

                          away_yellowCards  home_redCards  away_redCards  \
home                year                                                   
Atlanta United FC   2017                28              3              6   
                    2018                39              2              5   
                    2019                46              1              3   
  

In [16]:
# Export the grouped dataframe to "all_players_grouped.csv"
csv_dir = "C:/Users/mamoo/UT_DataScience/UTA-VIRT-DATA-PT-02-2021-U-B/FinalProject/Resources/Clean_csv/"
matches_grouped.to_csv(f"{csv_dir}matches_grouped.csv")

In [43]:
mls_df.head(2)

Unnamed: 0      id                 home              away  year  \
0        1797  336076      Colorado Rapids  Columbus Crew SC  2012   
1        1798  336077  Vancouver Whitecaps   Montreal Impact  2012   

   attendance                               venue game_status  shootout  \
0     14746.0  Dick's Sporting Goods Park, Denver          FT     False   
1     21000.0                 BC Place, Vancouver          FT     False   

   home_score  away_score  home_possessionPct  away_possessionPct  \
0           2           0                0.53                0.47   
1           2           0                0.50                0.50   

   home_foulsCommitted  away_foulsCommitted  home_yellowCards  \
0                   11                   14                 1   
1                   13                   26                 1   

   away_yellowCards  home_redCards  away_redCards  home_offsides  \
0                 3              0              0              2   
1                 2              0              0              2   

   away_offsides  home_wonCorners  away_wonCorners  home_saves  away_saves  \
0              4                8                4           3           3   
1              2                4                4           7           3   

  Outcome  
0     Win  
1     Win

## Export the DataFrame to PostgreSQL Database

In [39]:
#Add the code to create the connection to the PostgreSQL database.
# For our local server, the connection string will be as follows:
db_link = f"postgres://postgres:{db_password}@127.0.0.1:5432/major_league_soccer"    

In [40]:
# Create the database engine 
engine = create_engine(db_link)

In [41]:
# add the mls_df DataFrame to a SQL database.
mls_df.to_sql(name='matches_data', con=engine, if_exists='replace')
    

In [42]:
# add the matches_grouped DataFrame to a SQL database.
matches_grouped.to_sql(name='matches_grouped', con=engine, if_exists='replace')